scripts/solve_network.py

```diff
@@ -314,78 +319,72 @@ def add_chp_constraints(n):
+        link_p_nom = n.model["Link-p_nom"]
 
         # ratio of output heat to electricity set by p_nom_ratio
-        lhs = linexpr(
-            (
-                n.links.loc[electric_ext, "efficiency"]
-                * n.links.loc[electric_ext, "p_nom_ratio"],
-                link_p_nom[electric_ext],
-            ),
-            (-n.links.loc[heat_ext, "efficiency"].values, link_p_nom[heat_ext].values),
-        )
+        lhs = (
+            n.links.loc[electric_ext, "efficiency"]
+            * n.links.loc[electric_ext, "p_nom_ratio"]
+            * link_p_nom.loc[electric_ext]
+        ) + (-n.links.loc[heat_ext, "efficiency"] * link_p_nom.loc[heat_ext])
 
-        define_constraints(n, lhs, "=", 0, "chplink", "fix_p_nom_ratio")
+        n.model.add_constraints(lhs == 0, name="chplink-fix_p_nom_ratio")
 
         # top_iso_fuel_line for extendable
-        lhs = linexpr(
-            (1, link_p[heat_ext]),
-            (1, link_p[electric_ext].values),
-            (-1, link_p_nom[electric_ext].values),
+        lhs = (
+            (1 * link_p.loc[heat_ext])
+            + (1 * link_p.loc[electric_ext])
+            + (-1 * link_p_nom.loc[electric_ext])
         )
 
-        define_constraints(n, lhs, "<=", 0, "chplink", "top_iso_fuel_line_ext")
+        n.model.add_contraints(lhs <= 0, name="chplink-top_iso_fuel_line_ext")
 
     if not electric_fix.empty:
         # top_iso_fuel_line for fixed
-        lhs = linexpr((1, link_p[heat_fix]), (1, link_p[electric_fix].values))
+        lhs = (1 * link_p[heat_fix]) + (1 * link_p.loc[electric_fix])
 
-        rhs = n.links.loc[electric_fix, "p_nom"].values
+        rhs = n.links.loc[electric_fix, "p_nom"]
 
-        define_constraints(n, lhs, "<=", rhs, "chplink", "top_iso_fuel_line_fix")
+        n.model.add_constraints(lhs <= rhs, name="chplink-top_iso_fuel_line_fix")
 
     if not electric.empty:
         # backpressure
-        lhs = linexpr(
-            (
-                n.links.loc[electric, "c_b"].values * n.links.loc[heat, "efficiency"],
-                link_p[heat],
-            ),
-            (-n.links.loc[electric, "efficiency"].values, link_p[electric].values),
-        )
+        lhs = (
+            n.links.loc[electric, "c_b"].values
+            * n.links.loc[heat, "efficiency"]
+            * link_p.loc[heat]
+        ) + (-n.links.loc[electric, "efficiency"].values * link_p.loc[electric])
 
-        define_constraints(n, lhs, "<=", 0, "chplink", "backpressure")
+        n.model.add_constraints(lhs <= 0, name="chplink-backpressure")
 
 
 def add_co2_sequestration_limit(n, sns):
     co2_stores = n.stores.loc[n.stores.carrier == "co2 stored"].index
 
-    if co2_stores.empty or ("Store", "e") not in n.variables.index:
+    if co2_stores.empty:  # or ("Store", "e") not in n.variables.index:
         return
 
-    vars_final_co2_stored = get_var(n, "Store", "e").loc[sns[-1], co2_stores]
+    vars_final_co2_stored = n.model["Store-e"].loc[sns[-1], co2_stores]

```

Comment:
As a general comment, indexing in `linopy` may be a bit delicate. My observation is that there may be no one-to-one correspondence with old functions. Some testing may be needed, especially there is no exact match with PyPSA-Eur.

Reply:
Do you have any ideas on how we can test the functionalities?

from my end, I think we can test the original functionalities alongside the new ones, and compare the results.